app/assets/javascripts/code_listing/code_listing.ts

```diff
@@ -1,278 +1,348 @@
-        this.diffSwitchPrefix = document.querySelector("#diff-switch-prefix");
-
-        const showAllListener = (): void => {
-            this.showAllAnnotations();
-            this.annotationsWereHidden?.remove();
-        };
+    // /////////////////////////////////////////////////////////////////////////
+    // Highlighting ////////////////////////////////////////////////////////////
+    // /////////////////////////////////////////////////////////////////////////
 
-        this.showAllButton.addEventListener("click", () => showAllListener());
-        this.hideAllButton.addEventListener("click", () => this.hideAllAnnotations());
+    clearHighlights(): void {
+        const markedAnnotations = this.table.querySelectorAll(`tr.lineno.${this.markingClass}`);
+        markedAnnotations.forEach(markedAnnotation => {
+            markedAnnotation.classList.remove(this.markingClass);
+        });
+    }
 
-        this.showOnlyErrorButton.addEventListener("click", () => this.compressAnnotations());
+    highlightLine(lineNr: number, scrollToLine = false): void {
+        const toMarkAnnotationRow = this.table.querySelector(`tr.lineno#line-${lineNr}`);
+        toMarkAnnotationRow.classList.add(this.markingClass);
+        if (scrollToLine) {
+            toMarkAnnotationRow.scrollIntoView({ block: "center" });
+        }
     }
 
-    private setToggleButtonsVisibility(): void {
-        if (this.annotations.length > 0) {
-            this.showAllButton.classList.remove("hide");
-            this.hideAllButton.classList.remove("hide");
-            this.diffSwitchPrefix.classList.remove("hide");
+    // /////////////////////////////////////////////////////////////////////////
+    // Annotation management ///////////////////////////////////////////////////
+    // /////////////////////////////////////////////////////////////////////////
 
-            if (this.annotations.some(a => a.type === "error" || a.type === "user")) {
-                this.showOnlyErrorButton.classList.remove("hide");
-                this.annotationsWereHidden.classList.remove("hide");
-            }
+    private addAnnotation(annotation: Annotation): void {
+        const line = annotation.global ? 0 : annotation.line;
 
-            const errorCount = this.annotations.filter(a => a.type !== "error" && a.type !== "user").length;
-            if (errorCount > 0) {
-                const nonErrorAnnotationCount = this.createHiddenMessage(errorCount);
-                this.annotationsWereHidden.innerHTML = "";
-                this.annotationsWereHidden.appendChild(nonErrorAnnotationCount);
-            }
+        if (!this.annotations.has(line)) {
+            this.annotations.set(line, []);
+        }
+
+        // Add the annotation in the map.
+        const lineAnnotations = this.annotations.get(line);
+        lineAnnotations.push(annotation);
+        this.annotations.set(line, lineAnnotations);
```

Comment:
```suggestion
        this.annotations.get(line).push(annotation);
```
Isn't this equivalent?